### DISCUSSION TITLE
Caret position change event support

### DISCUSSION THREAD

**Discussion by HuangZhuo** [Taitava/obsidian-shellcommands#284] [Ideas - done & released]
I have a idea to use 「git blame」 to show current line change log, just like gitlens did.

**Comment by HuangZhuo** [comment]
a 「Every n secs」event could solve this, but a new event could do it better

**Comment by Taitava** [comment]
Thank you for your suggestion, it makes perfectly sense! 🙂 

It's actually listed [here](https://github.com/Taitava/obsidian-shellcommands/discussions/60#discussion-3602335), but anyway good to have a new discussion for it. 👍 
> ## Editor
> 
>  * _When the caret is moved_. Note that this can happen very rapidly, so the shell command should not be anything heavy.

## Quick trial

I tried now to quickly find out how to create this kind of event, but didn't succeed yet. I tried to examine the following event provided by Obsidian's API:
https://github.com/obsidianmd/obsidian-api/blob/902badd38ba907689f0917d7b193f7e33d1284fe/obsidian.d.ts#L4143-L4146

I don't have a clue about in which situation the event is fired - I was not able to fire it at all. So looks like I will not use that particular event for this, but don't know yet.

## Next thing to try

 Next I'll try the following from CodeMirror's API (`index.d.ts`):
```typescript
/** Will be fired when the cursor or selection moves, or any change is made to the editor content. */
on(eventName: 'cursorActivity', handler: (instance: CodeMirror.Editor) => void): void;
```

I just don't know yet how to hook into CodeMirror's events. Will try to figure out some time.

**Reply by HuangZhuo** [reply]
Thanks for your explanation! As you said, this is a rapid event, so I have an idea:
1. a line change event.
2. when caret line changes, dont fire the event instantly, instead, use a timer(should be a short period. like 0.5 secs) to fire the event later, when new event comes, reset the timer and save latest event.

I'm not familiar with plugin development, so the advice may not be appropriate.

**Reply by Taitava** [reply]
It's a good idea, actually something that would suit for many other events, too. The feature could be called as _threshold_.

It could be a shell command level setting:
 - _Threshold duration_: in milliseconds, e.g. `500` would be 0.5 seconds like you suggested. If `0`, threshold is not used.

I still need to think about if the threshold should only apply to events, or should it apply to executing shell commands via Obsidian's command palette/hotkeys and [Shell command URI](https://publish.obsidian.md/shellcommands/Basic+usage/Shell+commands+URI), too. I.e. prevent also manual execution if an event had just executed the shell command?

Edit: Also need to add an option defining whether the shell command execution should happen `early` or `late`:
 - Executing `early` first executes the shell command and then waits for the threshold duration. If any execution attempts occur during the waiting, they will be cancelled silently.
 - Executing `late` waits until the threshold is passed, and then executes the shell command. This is better if you need to ensure that also the last event occurrence will execute the shell command. Otherwise the final event occurring time might go unnoticed.
 - **OR**: Maybe consider if these could be combined: Execute immediately, then wait and see if further execution attempts are made, and if yes, do a re-execution after the threshold duration is over. No setting would be needed. I need to think about this.

**Reply by HuangZhuo** [reply]
- *Threshold* is designed to solve performance issues and avoid impacting the input experience when executed at high frequencies.

- Manual execution is usually not very frequent, and more important, manual execution **always** expects a result.

- This is a demo, the blame info always show lately, It may be more desirable to always execute the last event in this case.

![demo](https://user-images.githubusercontent.com/5159053/203676867-d3096c75-1db3-4598-8799-3bdbddf7d30e.gif)

**Reply by Taitava** [reply]
> * _Threshold_ is designed to solve performance issues and avoid impacting the input experience when executed at high frequencies.
> 
>  * Manual execution is usually not very frequent, and more important, manual execution **always** expects a result.

Yea. I was thinking more about an edge case where an automatic event has just executed a shell command, and then user executes it manually. I was wondering if the latter execution should be prevented, but I do agree it's not needed.

>  * This is a demo, the blame info always show lately, It may be more desirable to always execute the last event in this case.

Yes, in this case the last event should always execute. 👍

**Reply by Taitava** [reply]
@HuangZhuo regarding [your old comment above](https://github.com/Taitava/obsidian-shellcommands/discussions/284#discussioncomment-4144233):
> 1. a line change event.

Wow, I just noticed this in your message over a year later 🙂 . It could make perfect sense to have a mode setting in the _Caret moved_ event, that could limit when to fire the event:
- Only when line changes
- Only when column changes
- In both cases (default)

This would be an excellent addition to the event! Too bad, the event is not completed still. I'll need to look into it again, but I can't promise when I can do it.

**Reply by HuangZhuo** [reply]
Thank you, I'm looking forward to it.

**Comment by Taitava** [comment]
## Finally some progress

@HuangZhuo I have now finally been able to get the _Caret moved in editor_ event working! 🙂 Obsidian uses [CodeMirror 6](https://codemirror.net/) as its core editor component, and I needed to dig throught different guides and websites to learn how to write extensions for CodeMirror 6. A good thing is that now it might be a bit easier for me to develop other CodeMirror related events, too, if I ever need to.

There is still some work left for me to do, e.g. clean up the code and commit it to git, and I also need to find a way to detect if the caret moved to another line, or just to another column in the same line. I also need to test it more to find out if there are any problems with it. I just got it working today, so haven't had much time to try different things.

I'm planning to release this in SC `0.23.0` - so not in the next version `0.22.0`, but after that. 🙂

**Comment by Taitava** [comment]
## Variable idea: `{{event_old_caret_position}}`

If ever needed, I could create variables like `{{event_old_caret_position:line}}` and `{{event_old_caret_position:column}}` (similar to [`{{caret_position}}`](https://publish.obsidian.md/shellcommands/Variables/%7B%7Bcaret_position%7D%7D)) that could be used to know where the caret moved from to the current location.

I'm not implementing this now, but if a need arises, I'll do it later.

**Comment by Taitava** [comment]
@HuangZhuo 
> I have a idea to use 「git blame」 to show current line change log, just like gitlens did.

This is now released in SC version `0.23.0`. And it happens to be exactly two years since you requested this (2022-11-09). Sorry for the long wait! 😅 

Please let me know if you have any feedback. 🙂

**Reply by HuangZhuo** [reply]
Much Appreciate!
I just tested the following command on my Mac:

```shell
git blame --date=relative -L {{caret_position:line}},{{caret_position:line}} {{file_path:relative}} | awk -F'[(|)]' '{print $2}'
```

It generally works fine for me.
I also noticed a feature called "prolong cooldown," which provides an experience just like GitLens in VS Code.

Time flies, best wishes.🤝
